File: python/linked-list/merge_lists.py
# Merge Two Sorted Lists

# You are given the heads of two sorted linked lists list1 and list2.

# Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.

# Return the head of the merged linked list.

# Implementing a List Node
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.next = None

# Implementing a Linked List
class LinkedList:
    def __init__(self, head) -> None:
        self.head = head
        self.tail = head
    
    def add(self, node):
        if not self.head:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = node
# create nodes list 1
node1 = Node(2)
node2 = Node(3)
node3 = Node(5)
node4 = Node(6)
node5 = Node(6)

linkedList = LinkedList(node1)
linkedList.add(node2)
linkedList.add(node3)
linkedList.add(node4)
linkedList.add(node5)

# create nodes list 2
node6 = Node(1)
node7 = Node(4)
node8 = Node(5)
node9 = Node(9)
node10 = Node(10)

linkedList2 = LinkedList(node6)
linkedList2.add(node7)
linkedList2.add(node8)
linkedList2.add(node9)
linkedList2.add(node10)

def mergeTwoLists(head1: Node, head2: Node) -> Node:
    # create a new node for the merged list
    merge_node = Node()
    # save a refrence for this node
    merge_node_head = merge_node
    # loop through every nodes in lists
    while head1 and head2:
        # check for the lowest val in node from list 1
        if head1.val <= head2.val:
            # add node to the merged list
            merge_node.next = head1
            # move to the next node in list1
            head1 = head1.next
        else:
            # add node to the merged list
            merge_node.next = head2
            # move to the next node in list2
            head2 = head2.next
        # move to the next node in merged list
        merge_node = merge_node.next
    # add all the remaining nodes from list1 or list2
    merge_node.next = head1 or head2
    # return the head of the linked list
    return merge_node_head.next

File: javascript/linked-list/merge-lists.js
// Merge Two Sorted Lists

// You are given the heads of two sorted linked lists list1 and list2.

// Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.

// Return the head of the merged linked list.

// impelemnting node 
class ListNode {
    constructor(val) {
        this.val = val
        this.next = null
    }
}

// Implementing a Linked List
class LinkedList {
    constructor(head = null) {
        this.head = head
        this.tail = head
    }

    add(node) {
        if (!this.head) {
            this.head = node
            this.tail = node
        } else {
            this.tail.next = node
            this.tail = node
        }
    }
}
// nodes list 1
let node1 = new ListNode(1)
let node2 = new ListNode(2)
let node3 = new ListNode(4)
let node4 = new ListNode(6)
let node5 = new ListNode(7)

let list1 = new LinkedList(node1)
list1.add(node2)
list1.add(node3)
list1.add(node4)
list1.add(node5)
// nodes list 2
let node6 = new ListNode(1)
let node7 = new ListNode(3)
let node8 = new ListNode(4)
let node9 = new ListNode(5)
let node10 = new ListNode(8)

let list2 = new LinkedList(node6)
list2.add(node7)
list2.add(node8)
list2.add(node9)
list2.add(node10)

function mergeTwoLists(head1, head2) {
    // create a node for the new merged list
    let mergedNode = new ListNode()
    // save a referance of the list head
    const mergedNodeHead = mergedNode
    // loop through nodes lists
    while (head1 && head2) {
        if (head1.val <= head2.val) {
            // link the nodes with the lower node value
            mergedNode.next = head1
            // move to the next node
            head1 = head1.next
        } else {
            // link the nodes with the lower node value
            mergedNode.next = head2
            // move to the next node
            head2 = head2.next
        }
        // move to the next node in merged list
        mergedNode = mergedNode.next
    }
    // move all the remaining nodes from one of the list
    mergedNode.next = head1 || head2
    // return the head of the new list
    return mergedNodeHead.next
}

// testing
console.log(mergeTwoLists(list1.head, list2.head))
// expected output: ListNode{val:1, next: ListNode{val:1, [...]}}